File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::collections::HashMap;
        use std::cmp::max;
        
        let mut map: HashMap<char, i32> = HashMap::new();
        let mut i = 0;
        let mut longest = 0;
        
        for (j, char) in s.chars().enumerate(){
            if let Some(x) = map.get(&char){
                i = max(*map.get(&char).unwrap(), i);
            }
            longest = max(longest, j as i32-i+1);
            map.insert(char, j as i32 + 1);
        }
        
        longest
        
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
from collections import Counter
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        chars = Counter()
        
        left = 0
        right = 0
        longest = 0
        
        while right < len(s):
            chars[s[right]] += 1
            while chars[s[right]] > 1:
                chars[s[left]] -=1
                left+=1
            
            longest = max(longest, right-left+1)
            right+=1
        
        return longest